File: frontend/src/services/papeleraService.js
// =============================================
// 🔹 Servicio: Papelera
// Archivo: src/services/papeleraService.js
// =============================================
import api from "./api";

// ===============================
// 🔸 Obtener Promesas en Papelera
// ===============================
export const obtenerPromesasPapelera = async (usuarioId) => {
  try {
    const response = await api.get(`/promesas/listar/${usuarioId}`);
    // Filtra las que estén inactivas (activo == false o 0)
    return response.data.filter((p) => !p.activo);
  } catch (err) {
    console.error("[servicio] obtenerPromesasPapelera error:", err.response?.data || err);
    throw err.response?.data || { message: err.message || "Error al obtener promesas en papelera" };
  }
};

// ===============================
// 🔸 Obtener Notas de Diario en Papelera
// ===============================
export const obtenerNotasPapelera = async (usuarioId) => {
  try {
    const response = await api.get(`/diario/listar/${usuarioId}`);
    return response.data.filter((n) => !n.activo);
  } catch (err) {
    console.error("[servicio] obtenerNotasPapelera error:", err.response?.data || err);
    throw err.response?.data || { message: err.message || "Error al obtener notas en papelera" };
  }
};

// ===============================
// 🔸 Obtener Motivaciones en Papelera
// ===============================
export const obtenerMotivacionesPapelera = async (usuarioId) => {
  try {
    const response = await api.get(`/motivaciones/listar/${usuarioId}`);
    return response.data.filter((m) => !m.activo);
  } catch (err) {
    console.error("[servicio] obtenerMotivacionesPapelera error:", err.response?.data || err);
    throw err.response?.data || { message: err.message || "Error al obtener motivaciones" };
  }
};

// ===============================
// 🔸 Restaurar elemento (poner activo = true)
// ===============================
export const restaurarElemento = async (tipo, id) => {
  try {
    let response;
    switch (tipo) {
      case "Promesas":
        response = await api.put(`/promesas/${id}/estado?estado=true`);
        break;
      case "Entradas de Diario":
        response = await api.put(`/diario/${id}/estado?estado=true`);
        break;
      case "Motivaciones":
        response = await api.put(`/motivaciones/${id}/estado?estado=true`);
        break;
      default:
        throw new Error("Tipo de elemento no reconocido");
    }
    return response.data;
  } catch (err) {
    console.error("[servicio] restaurarElemento error:", err.response?.data || err);
    throw err.response?.data || { message: err.message || "Error al restaurar elemento" };
  }
};

// ===============================
// 🔸 Eliminar definitivamente
// ===============================
export const eliminarDefinitivo = async (tipo, id) => {
  try {
    let response;
    switch (tipo) {
      case "Promesas":
        response = await api.delete(`/promesas/${id}`);
        break;
      case "Entradas de Diario":
        response = await api.delete(`/diario/${id}`);
        break;
      case "Motivaciones":
        response = await api.delete(`/motivaciones/${id}`);
        break;
      default:
        throw new Error("Tipo de elemento no reconocido");
    }
    return response.data;
  } catch (err) {
    console.error("[servicio] eliminarDefinitivo error:", err.response?.data || err);
    throw err.response?.data || { message: err.message || "Error al eliminar elemento" };
  }
};

File: backend/app/services/papelera_service.py
from fastapi import HTTPException
from sqlalchemy.orm import Session
from app.models.promesa import Promesa
from app.models.notadiario import NotaDiario
from app.models.motivacion import Motivacion
from app.models.categoria import Categoria


class PapeleraService:
    # ===============================
    # 🔹 PROMESAS INACTIVAS
    # ===============================
    @staticmethod
    def obtener_promesas_papelera(db: Session, usuario_id: int):
        return db.query(Promesa).filter(
            Promesa.usuario_id == usuario_id,
            Promesa.activo == False
        ).all()

    # ===============================
    # 🔹 NOTAS DE DIARIO INACTIVAS
    # ===============================
    @staticmethod
    def obtener_notas_papelera(db: Session, usuario_id: int):
        return db.query(NotaDiario).filter(
            NotaDiario.usuario_id == usuario_id,
            NotaDiario.activo == False
        ).all()

    # ===============================
    # 🔹 MOTIVACIONES INACTIVAS
    # ===============================
    @staticmethod
    def obtener_motivaciones_papelera(db: Session, usuario_id: int):
        return db.query(Motivacion).filter(
            Motivacion.usuario_id == usuario_id,
            Motivacion.activo == False
        ).all()

    # ===============================
    # 🔹 CATEGORÍAS INACTIVAS
    # ===============================
    @staticmethod
    def obtener_categorias_papelera(db: Session, usuario_id: int):
        return db.query(Categoria).filter(
            Categoria.usuario_id == usuario_id,
            Categoria.activo == False
        ).all()

    # ===============================
    # 🔹 RESTAURAR ELEMENTO (activo=True)
    # ===============================
    @staticmethod
    def restaurar_elemento(db: Session, tipo: str, id: int):
        modelos = {
            "promesa": Promesa,
            "diario": NotaDiario,
            "motivacion": Motivacion,
            "categoria": Categoria,
        }

        modelo = modelos.get(tipo)
        if not modelo:
            raise HTTPException(status_code=400, detail="Tipo inválido")

        elemento = db.query(modelo).filter(modelo.id == id).first()
        if not elemento:
            raise HTTPException(status_code=404, detail=f"{tipo.capitalize()} no encontrada")

        # ✅ Restaurar la categoría y sus motivaciones asociadas
        if tipo == "categoria":
            elemento.activo = True
            db.commit()
            db.refresh(elemento)

            # Reactivar motivaciones asociadas
            motivaciones_afectadas = db.query(Motivacion).filter(
                Motivacion.categoria_id == id
            ).update({"activo": True})
            db.commit()

            return {
                "mensaje": f"Categoría '{elemento.nombre}' restaurada con {motivaciones_afectadas} motivaciones asociadas."
            }

        # ✅ Para el resto de tipos
        elemento.activo = True
        db.commit()
        db.refresh(elemento)
        return {"mensaje": f"{tipo.capitalize()} restaurada correctamente"}

    # ===============================
    # 🔹 ELIMINAR DEFINITIVAMENTE
    # ===============================
    @staticmethod
    def eliminar_definitivo(db: Session, tipo: str, id: int):
        modelos = {
            "promesa": Promesa,
            "diario": NotaDiario,
            "motivacion": Motivacion,
            "categoria": Categoria,
        }

        modelo = modelos.get(tipo)
        if not modelo:
            raise HTTPException(status_code=400, detail="Tipo inválido")

        elemento = db.query(modelo).filter(modelo.id == id).first()
        if not elemento:
            raise HTTPException(status_code=404, detail=f"{tipo.capitalize()} no encontrada")

        db.delete(elemento)
        db.commit()
        return {"mensaje": f"{tipo.capitalize()} eliminada definitivamente"}
